### PR TITLE
(DOCS-7865)(App Builder) JS Expressions

### PR DESCRIPTION
**Pull request by hestonhoffman**
### Merge instructions

- [x] Please merge after reviewing

### CODE REVIEW

**Review by kc-beard — APPROVED** [review]

File: content/en/service_management/app_builder/expressions.md [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+  tag: "Documentation"
+  text: "Build Apps"
+- link: "/service_management/app_builder/components/"
+  tag: "Documentation"
+  text: "Components"
+---
+
+You can use JavaScript (JS) expressions anywhere in App Builder to create custom interactions between the different parts of your app. As you type an expression, App Builder offers autocomplete suggestions based on the existing queries and components in your app. Click on an autocomplete suggestion to use it in your expression, or use the arrow keys on your keyboard and make a selection with the Enter key.
+
+Some fields, like the [post-query transformation][1], display a code editor by default and accept plain JS. In all other fields, enclose your JS expressions in `${}`. For example, to interpolate the values of two text input components named `textInput0` and `textInput1` into the **Content** property of a text component (and add an exclamation mark), use the expression `${textInput0.value} ${textInput1.value}!`.
```
Comment (neko-dd) [inline content/en/service_management/app_builder/expressions.md:15]:
```suggestion
Some fields, like [post-query transformation][1], display a code editor by default and accept plain JS. In all other fields, enclose your JS expressions in `${}`. For example, to interpolate the values of two text input components named `textInput0` and `textInput1` into the **Content** property of a text component (and add an exclamation mark), use the expression `${textInput0.value} ${textInput1.value}!`.
```

File: content/en/service_management/app_builder/expressions.md [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+{{< img src="service_management/app_builder/interpolation.png" alt="The text component fills with the words 'Hello' and 'World', each interpolated from a text input component value" style="width:100%;" >}}
+
+App Builder accepts standard vanilla JavaScript syntax, however:
```
Comment (neko-dd) [inline content/en/service_management/app_builder/expressions.md:19]:
```suggestion
App Builder accepts standard vanilla JavaScript syntax, with the following caveats:
```

File: content/en/service_management/app_builder/expressions.md [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+- The result of the expression must match the result expected by the component or query property. For example, the text component's **Is Visible** property expects a Boolean. To find out what type of data a component property expects, see [View component properties](#view-component-properties).
+- Your code has read-only access to the app state, but App Builder executes the code in a sandboxed environment with no access to the Document Object Model (DOM) or browser APIs.
+
+## View component properties
+
+Before you create an expression, it's helpful to know the available properties and defaults or current values for the component you want to interact with.
+
+You can view the available properties and values for a component using:
```
Comment (neko-dd) [inline content/en/service_management/app_builder/expressions.md:27]:
```suggestion
You can view the available properties and values for a component in the following ways:
```

File: content/en/service_management/app_builder/expressions.md [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+Some fields, like the [post-query transformation][1], display a code editor by default and accept plain JS. In all other fields, enclose your JS expressions in `${}`. For example, to interpolate the values of two text input components named `textInput0` and `textInput1` into the **Content** property of a text component (and add an exclamation mark), use the expression `${textInput0.value} ${textInput1.value}!`.
+
+{{< img src="service_management/app_builder/interpolation.png" alt="The text component fills with the words 'Hello' and 'World', each interpolated from a text input component value" style="width:100%;" >}}
+
+App Builder accepts standard vanilla JavaScript syntax, however:
+- The result of the expression must match the result expected by the component or query property. For example, the text component's **Is Visible** property expects a Boolean. To find out what type of data a component property expects, see [View component properties](#view-component-properties).
+- Your code has read-only access to the app state, but App Builder executes the code in a sandboxed environment with no access to the Document Object Model (DOM) or browser APIs.
+
```
Comment (neko-dd) [inline content/en/service_management/app_builder/expressions.md:22]: I see that we've dropped the section on the `?.` operator, but won't it be confusing to users when that shows up in autocompleted strings but we never address or explain it?

File: content/en/service_management/app_builder/expressions.md [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+To access the **Admin Console**:
+1. Click on the cog (**Settings**) icon and select **Admin Console**.
+1. Click **Data**.
+
+{{< img src="service_management/app_builder/admin-console.png" alt="The App State section in App Properties" style="width:100%;" >}}
+{{% /collapse-content %}}
+
+## Custom component interactions
+
+Most UI components provide built-in options such as toggles and text alignment that cover basic app usage. To add a custom interaction to a component, click the code editor symbol (**</>**) and enter a JS expression.
```
Comment (neko-dd) [inline content/en/service_management/app_builder/expressions.md:58]:
```suggestion
Most UI components provide built-in options, such as toggles and text alignment, that cover basic app usage. To add a custom interaction to a component, click the code editor symbol (**</>**) and enter a JS expression.
```

File: content/en/service_management/app_builder/expressions.md [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+1. Click the code editor symbol (**</>**) next to the **Is Disabled** property.
+1. Add the expression `${!text0.isVisible}`.
+
+The text component is invisible and the button is disabled unless both text input fields have content.
+
+{{< img src="service_management/app_builder/is-disabled.png" alt="The text component is invisible and the button is disabled unless both text input fields have content." style="width:100%;" >}}
+
+### Disable a component while loading
+
+Another common use case is disabling a component while a query is in a loading state. In the [EC2 Instance Manager blueprint][3], the `instanceType` Select component is disabled while the `listInstances` query is loading. To accomplish this, the **Is Disabled** property uses the expression `${listInstances.isLoading}`.
```
Comment (neko-dd) [inline content/en/service_management/app_builder/expressions.md:81]:
```suggestion
Another common use case is disabling a component while a query is in a loading state. In the [EC2 Instance Manager blueprint][3], the `instanceType` select component is disabled while the `listInstances` query is loading. To accomplish this, the **Is Disabled** property uses the expression `${listInstances.isLoading}`.
```

File: content/en/service_management/app_builder/expressions.md [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+{{< img src="service_management/app_builder/inspect-data.png" alt="The App State section in App Properties" style="width:100%;" >}}
+{{% /collapse-content %}}
+
+{{% collapse-content title="Admin Console" level="h3" %}}
+To access the **Admin Console**:
+1. Click on the cog (**Settings**) icon and select **Admin Console**.
+1. Click **Data**.
+
+{{< img src="service_management/app_builder/admin-console.png" alt="The App State section in App Properties" style="width:100%;" >}}
+{{% /collapse-content %}}
```
Comment (neko-dd) [inline content/en/service_management/app_builder/expressions.md:54]: IMO, these screenshots aren't adding much to the user experience, but potentially add a lot to maintenance overhead.

File: content/en/service_management/app_builder/expressions.md [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+The text component is invisible and the button is disabled unless both text input fields have content.
+
+{{< img src="service_management/app_builder/is-disabled.png" alt="The text component is invisible and the button is disabled unless both text input fields have content." style="width:100%;" >}}
+
+### Disable a component while loading
+
+Another common use case is disabling a component while a query is in a loading state. In the [EC2 Instance Manager blueprint][3], the `instanceType` Select component is disabled while the `listInstances` query is loading. To accomplish this, the **Is Disabled** property uses the expression `${listInstances.isLoading}`.
+
+{{< img src="service_management/app_builder/isloading.png" alt="The 'instanceType' Select component is disabled while the 'listInstances' query is loading." style="width:100%;" >}}
```
Comment (neko-dd) [inline content/en/service_management/app_builder/expressions.md:83]: This screenshot also feels like it doesn't add much to the user experience.

File: content/en/service_management/app_builder/expressions.md [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+
+## Custom query interactions
+
+Similar to components, you can use JS expressions to alter your queries based on user interaction.
+
+### Filter query results on user input
+
+The [PagerDuty On-call Manager blueprint][4] filters the result of the `listSchedules` query based on input from the user. The user selects a team and user from the `team` and `user` **Select** components.
```
Comment (neko-dd) [inline content/en/service_management/app_builder/expressions.md:91]:
```suggestion
The [PagerDuty On-call Manager blueprint][4] filters the result of the `listSchedules` query based on input from the user. The user selects a team and user from the `team` and `user` select components.
```

File: content/en/service_management/app_builder/expressions.md [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+{{< /code-block >}}
+
+Setting the query's **Run Settings** to **Auto** allows the query to run each time a user changes a value in the `team` or `user` components.
+
+## Further reading
+
+{{< partial name="whats-next/whats-next.html" >}}
+
+[1]: /service_management/app_builder/build/#post-query-transformation
+[2]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining
```
Comment (neko-dd) [inline content/en/service_management/app_builder/expressions.md:119]: IFF we leave out the section above, you no longer need this link

File: config/_default/menus/main.en.yaml
Comment (neko-dd) [inline config/_default/menus/main.en.yaml:1]: Looks like you need to pull in some changes to this file

File: content/en/service_management/app_builder/expressions.md [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+        teams: s.teams.map(u => u.summary)
+    }
+}).filter(s => {
+
+        const matchesName = !name.value.length ? true : s.name.toLowerCase().includes(name.value.toLowerCase());
+        const matchesTeam = team.value === 'Any Team' ? true : s.teams.includes(team.value);
+        const matchesUser = user.value === 'Any User' ? true : s.users.includes(user.value);
+
+        return matchesName && matchesUser && matchesTeam ;
+    }) || []
```
Comment (neko-dd) [inline content/en/service_management/app_builder/expressions.md:109]: I'm just curious what the `|| []` bit at the end means/does

File: content/en/service_management/app_builder/expressions.md (hunk 4 above)
Comment (hestonhoffman) [reply content/en/service_management/app_builder/expressions.md:22]: This is one of those hills.

File: content/en/service_management/app_builder/expressions.md (hunk 7 above)
Comment (hestonhoffman) [reply content/en/service_management/app_builder/expressions.md:54]: Honestly, the UI right now can be a little unintuitive. I wanted to give at least a couple of examples of the kind of places you can enter expressions. Perhaps if the UI is improved later, the maintenance could be to remove the images.

File: content/en/service_management/app_builder/expressions.md (hunk 11 above)
Comment (hestonhoffman) [reply content/en/service_management/app_builder/expressions.md:109]: I believe it assigns an empty array if the previous expression returns a falsy/NaN etc.

File: content/en/service_management/app_builder/expressions.md (hunk 10 above)
Comment (hestonhoffman) [inline content/en/service_management/app_builder/expressions.md:119]:
```suggestion
```

File: content/en/service_management/app_builder/expressions.md (hunk 11 above)
Comment (neko-dd) [reply content/en/service_management/app_builder/expressions.md:109]: Neat, thanks

**Review by neko-dd — APPROVED** [review]
LGTM, thanks

**Pull request merged**